### DISCUSSION TITLE
Scoop vs winget

### DISCUSSION THREAD

**Discussion by wizdude** [ScoopInstaller/Scoop#4777] [General]
i'm just curious. with Microsoft's winget getting a lot of traction these days, what the main selling point or feature differences between Scoop and winget?

for anyone using this to automate machine installs and the like, there is a lot of time invested - regardless which tool/solution you use.

has Scoop got a bigger repository?

i've just started investing time in winget for SOE deployment. should I be spending time with Scoop instead?

i only heard about this project today.

many thanks :-)

cheers, Wiz!!

**Comment by niheaven** [comment]
There are many discussions regarding scoop and chocolatey, winget, etc., and you can see winget adopt many scoop features.

But nowadays, the main difference between scoop and winget is that winget is a installer manager, while scoop is a package manager.

tldr, winget just helps you find apps, and installs the app likes it usually is, to the location it's usually located; scoop tries to extract the app's installer and put the program's files in a user-wide central location, to avoid UAC, if possible.

**Reply by salim-b** [reply]
> But nowadays, the main difference between scoop and winget is that winget is a installer manager, while scoop is a package manager.

Amusingly, this quite contrasts with the following statement on the [Chocolatey and Winget Comparison](https://github.com/ScoopInstaller/Scoop/wiki/Chocolatey-and-Winget-Comparison) wiki page:

> **Simpler than packaging.** Scoop isn't a package manager, rather it reads plain JSON manifests that describe how to install a program and its dependencies.

(I guess that just proves terminology (*naming* in general) isn't easy. 😅)

**Comment by rashil2000** [comment]
You could think of winget as "the agent to download and run the installer". That's all, everything else is the same. It just invokes the GUI installer. All of the things related to installation - like installing to the default location, creating registry keys, adding binaries and shortcuts etc. is done by the installer itself, winget plays no role.

Scoop is completely different. It offers full control, extracts and installs to a common (and configurable) location, manages persistent data etc. It is mostly self-contained - no part outside the Scoop directory is modified - and it does not touch the registry and any machine-specific settings.

That said, because of the way Scoop handles installations, the out-of-the-box experience can differ for some applications, when compared to running their official installers.

**Reply by ghost** [reply]
Hi,
if it doesn't touch the system registry, how programs add their keys to the registry then? lots of programs rely on certain registry keys present on system to work, how does Scoop handle that? thanks

**Reply by rashil2000** [reply]
Scoop itself does not touch the registry, but an installed program (during its runtime) can still interface with it.

**Reply by rasa** [reply]
Scoop doesn't touch the registry by design, but many programs provide scripts that the user can run to update the registry, if they want.

**Reply by ghost** [reply]
I see, thank you, so does it mean I can install any program I want using Scoop from Scoop buckets, then move Scoop directory to another computer, use [this method](https://github.com/ScoopInstaller/Scoop/issues/2894#issuecomment-447276027) to reinstall Scoop on new computer and start using programs without redownloading them? (I already tried that method and it worked, but need to know any program I install through Scoop will support that)

what I really want to know is that, does Scoop make programs portable? even those that heavily rely on system registry or install Windows services?

from what you say, I think you mean during program installations, they don't create registry keys, only when I run programs I install they have the chance to install registry keys and possiblity ask for elevation of privileges? 

I tried Sandboxie for a lot of programs but it was extremely disappointing all of the programs I tried failed with errors and only 1 of them barely worked, Sandboxie has so many compatibility issues. that's why I'm looking at Scoop as a better alternative, to make my programs portable.

thanks

**Reply by rashil2000** [reply]
> I see, thank you, so does it mean I can install any program I want using Scoop from Scoop buckets, then move Scoop directory to another computer, use [this method](https://github.com/ScoopInstaller/Scoop/issues/2894#issuecomment-447276027) to reinstall Scoop on new computer and start using programs without redownloading them? (I already tried that method and it worked, but need to know any program I install through Scoop will support that)

It should mostly work - but if an application creates files/registry entries specific to a machine, then you will have to reinstall that application.

> what I really want to know is that, does Scoop make programs portable? even those that heavily rely on system registry or install Windows services?

Scoop does not make programs portable. It _installs_ programs. Whether they are portable depends on the program developer.

> from what you say, I think you mean during program installations, they don't create registry keys, only when I run programs I install they have the chance to install registry keys and possiblity ask for elevation of privileges?

Precisely.

> I tried Sandboxie for a lot of programs but it was extremely disappointing all of the programs I tried failed with errors and only 1 of them barely worked, Sandboxie has so many compatibility issues. that's why I'm looking at Scoop as a better alternative, to make my programs portable.

Scoop and Sandboxie are different applications with different purposes. Scoop is a package manager, while Sandboxie is a way to run programs in an isolated way to prevent possible harm to a machine.

**Comment by joshjm** [comment]
IMO id love to see this breakdown added to the main `readme`, It's exactly what I need to know about scoop given that ive already used `winget`

**Reply by rashil2000** [reply]
I have added a link to this discussion in the Wiki: https://github.com/ScoopInstaller/Scoop/wiki/Chocolatey-and-Winget-Comparison

**Reply by Aragorn450** [reply]
> I have added a link to this discussion in the Wiki: https://github.com/ScoopInstaller/Scoop/wiki/Chocolatey-and-Winget-Comparison

You should also update the [Overview](https://github.com/ScoopInstaller/Scoop/wiki#overview) section of the Wiki Home as well as the [So What](https://github.com/ScoopInstaller/Scoop/wiki/So-What#but-i-already-use-x-why-should-i-use-scoop) page as both of their links to the Chocolatey comparison are broken.

**Reply by rashil2000** [reply]
Done!

**Comment by belgoros** [comment]
It works, in general, like a charm on a Windows laptop where I need admin rights to install the dev stuff. The only problem I hit was installing Elixir which is actually compiled to Erlang 22 but the scoop offers Erlang 25 (the latest) version only. It would be really great to be able to choose a version to install.

**Reply by rashil2000** [reply]
`scoop install erlang@22.0`

**Reply by belgoros** [reply]
Unfortunately, it didn't work:
```
Scoop was updated successfully!
WARN  Given version (22) does not match manifest (25.0.2)
WARN  Attempting to generate manifest for 'erlang' (22)
Autoupdating erlang
Downloading otp_win32_22.exe to compute hashes!
The remote server returned an error: (404) Not Found.
URL https://erlang.org/download/otp_win32_22.exe#/dl.7z is not valid
Could not install erlang@22
Select-CurrentVersion : Cannot bind argument to parameter 'AppName' because it is an empty string.
At C:\Users\scambour\scoop\apps\scoop\current\lib\core.ps1:236 char:54
+     return $null -ne (Select-CurrentVersion -AppName $app -Global:$gl ...
+                                                      ~~~~
    + CategoryInfo          : InvalidData: (:) [Select-CurrentVersion], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorEmptyStringNotAllowed,Select-CurrentVersion

You cannot call a method on a null-valued expression.
At C:\Users\scambour\scoop\apps\scoop\current\lib\manifest.ps1:28 char:5
+     $app = $app.TrimStart('/')
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

Couldn't find manifest for ''.
```
It would be really great to have the same option as **asdf** or **rvm** provide

**Reply by rashil2000** [reply]
You missed the `.0`

**Reply by belgoros** [reply]
My bad, nice catch! Thanks a lot, it worked like a charm 🚀 
Here is the output with some errors but elixir works (iex, mix commands, etc.):
```
Running post_install script...
Remove-Item : Cannot find path 'C:\Users\scambour\scoop\apps\erlang\current\Uninstall.exe.nsis' because it does not
exist.
At line:1 char:1
+ Remove-Item "$dir\`$PLUGINSDIR", "$dir\Uninstall.exe.nsis", "$dir\vcr ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\scambo...nstall.exe.nsis:String) [Remove-Item], ItemNotFoundEx
   ception
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand

'erlang' (22.0) was installed successfully!
'erlang' suggests installing 'extras/vcredist2013'.
```

**Reply by rashil2000** [reply]
Yeah, some errors are to be expected as the manifest is auto generated

**Comment by nbolton** [comment]
For situations where WinGet is not installed by default, Scoop is easier to install than WinGet because it doesn't have dependencies.

# Scoop
```
Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser;
Invoke-RestMethod -Uri https://get.scoop.sh | Invoke-Expression;
```

# Winget
```
$progressPreference = 'silentlyContinue'
Write-Information "Downloading WinGet and its dependencies..."
Invoke-WebRequest -Uri https://aka.ms/getwinget -OutFile Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle
Invoke-WebRequest -Uri https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx -OutFile Microsoft.VCLibs.x64.14.00.Desktop.appx
Invoke-WebRequest -Uri https://github.com/microsoft/microsoft-ui-xaml/releases/download/v2.8.6/Microsoft.UI.Xaml.2.8.x64.appx -OutFile Microsoft.UI.Xaml.2.8.x64.appx
Add-AppxPackage Microsoft.VCLibs.x64.14.00.Desktop.appx
Add-AppxPackage Microsoft.UI.Xaml.2.8.x64.appx
Add-AppxPackage Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle
```
Docs: https://learn.microsoft.com/en-us/windows/package-manager/winget/
